lib/_tls_wrap.js

```diff
@@ -382,34 +398,38 @@ TLSSocket.prototype._wrapHandle = function(wrap) {
   }
 
   // Wrap socket's handle
-  var context = options.secureContext ||
-                options.credentials ||
-                tls.createSecureContext(options);
-  res = tls_wrap.wrap(handle._externalStream,
-                      context.context,
-                      !!options.isServer);
+  const context = options.secureContext ||
+                  options.credentials ||
+                  tls.createSecureContext(options);
+  const res = tls_wrap.wrap(handle._externalStream,
+                            context.context,
+                            !!options.isServer);
   res._parent = handle;
   res._parentWrap = wrap;
   res._secureContext = context;
   res.reading = handle.reading;
+  this[kRes] = res;
```

Comment:
I do not think this is a good idea, is it possible to remove it without allocating any more closures?

Reply:
Not that I can tell. Doing it this way is more performant based on my testing but can revert.

Reply:
Can you please benchmark with and without? it does not seem to be needed, but I would be happy to be wrong.

Reply:
Sorry, what I mean is: I can't remove the closure without using `this[kRes]`. If we're ok with having a closure then this is fine. The performance difference is minor.

Reply:
Which closure?

Reply:
Lines 404-408 in the old code. `this.on('close')` As I said, it's not a huge deal. It makes a minor difference but really nothing major.

https://github.com/apapirovski/node/blame/7a953929fef5a7892d4697ed73e6fd314055beee/lib/_tls_wrap.js#L404-L408

Reply:
The cyclic reference still exists, as the closures below are closing over `res`. The actual solution is to move the whole `Object.defineProperty` to an outside function that is not passed `res`to. Then `this[kRes]` is correct.

Reply:
I might be going crazy here, sorry for dragging this on but...

```js
  this[kRes] = res;
  Object.defineProperty(handle, 'reading', {
    get: () => {
      return this[kRes].reading;
    },
    set: (value) => {
      this[kRes].reading = value;
    }
  });


  this.on('close', onSocketCloseDestroySSL);
```

Then we unset this[kRes] and res is no longer referenced by anything. I could be completely nuts though, been a long day.

Reply:
You would have to wait for the handle to be collected to be able to free `res`, while in this way it could be freed earlier. It was the spirit of https://github.com/apapirovski/node/blame/7a953929fef5a7892d4697ed73e6fd314055beee/lib/_tls_wrap.js#L404-L408, which makes the getter and setter refer to nothing, and it clear `res` from those closures as well.

The function call can likely be inlined by V8, so there should not be drop in performance but verify.

Reply:
I think this is the most controversial part here. As the diff is definitely not easy to read here I would suggest to remove this optimization for now and you could open another PR for that or at least have a individual commit for it. The latter would make the review a lot easier.

Reply:
I've updated it based on @mcollina's feedback. I think it should be good now?